action/selection/open_spiel/database.cc

```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
         } else if (use_hypopg && use_microservice) {
           query = absl::StrCat("EXPLAIN (FORMAT JSON) ", query);
           pqxx::result rset{txn.exec(query)};
-          ExplainMicroserviceCost cost{client, &rset};
+          ExplainMicroserviceCost cost{client, &rset, record_predictions};
           total_cost += cost.model_cost_ * work.num_arrivals_;
+          if (record_predictions) {
+            predicted_workloads.emplace_back(cost.model_cost_, cost.inference_results_.str());
+          }
         } else {
-          query = absl::StrCat("EXPLAIN (ANALYZE, BUFFERS) ", query);
-          pqxx::result rset{txn.exec(query)};
-          ExplainAnalyzeCost cost{&rset};
-          total_cost += cost.actual_total_time_ * work.num_arrivals_;
+          total_cost += GetExplainAnalyzeCostUs(txn, query).second * work.num_arrivals_;
+        }
+      }
+    }
+
+    if (use_hypopg && use_microservice && record_predictions) {
+      // Remove all hypothetical indexes
+      txn.exec("select hypopg_reset();");
+
+      // Start a sub-transaction. When this sub-transaction goes away, all the actions (which
+      // we assume are transactional) will be automatically rolled back.
+      pqxx::subtransaction subtxn(txn);
+      ApplyHistory(subtxn, false);
```

Comment:
This is where you would want something like

```suggestion
      CustomApplyHistory(subtxn, false);
```